### PR TITLE
fix(pkg): resolve symlinks from (install) action 

### PR DESCRIPTION
**Pull request by Alizter**
The reloctable compiler, like the other OCaml compilers, has symlinks in it's installation layout. These need to be resolved before we declare them inside the directory target or else the reloctable compiler cannot be cached. This work is tracked in #13229.

In the second commit we introduce a general mechanism to comb through the targets of the install layout for a package and resolve these symlinks by replacing them with a hardlink, falling back to a copy if that is not possible.

We add quite a few tests to exercise the numerous edge cases that come up.

- TODO describe tests a bit more

### CODE REVIEW

File: test/blackbox-tests/test-cases/pkg/install-symlinks/dir-symlink.t [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+--------------------------------------------------------------------------------
+
+Case 1: No .install file, directory symlink in %{lib}, engine rejects it.
+
+  $ make_lockpkg foo <<EOF
+  > (version 0.0.1)
+  > (build
+  >  (progn
+  >   (run mkdir -p %{lib}/%{pkg-self:name})
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/pkg/install-symlinks/dir-symlink.t]: Here's a suggestion to improve the readability:

Instead of adding a `run` for every single command, why not a single `bash` action with a multi line string? They were made for exactly this purpose.

Comment (Alizter) [reply test/blackbox-tests/test-cases/pkg/install-symlinks/dir-symlink.t]: We don't support `(bash)` in lock files.

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/pkg/install-symlinks/dir-symlink.t]: Okay. Use `sh`

Comment (Alizter) [reply test/blackbox-tests/test-cases/pkg/install-symlinks/dir-symlink.t]: Do you want them in a single `sh -c` or line by line?

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/pkg/install-symlinks/dir-symlink.t]: If it was line by line, I think it would defeat the purpose of the change. Since you're already using the shell, use it all the way and invoke it only once do all the work.

Comment (rgrinberg) [reply test/blackbox-tests/test-cases/pkg/install-symlinks/dir-symlink.t]: Similarly, in the actions where you don't use the shell (I think there's a couple?) you can leave things as is and use the dune language for actions.

File: src/dune_rules/pkg_rules.ml [hunk 2]
```diff
@@ -2011,6 +2043,14 @@ module Install_action = struct
...
         { Install_cookie.Gen.files; variables }
       in
+      (* Resolve symlinks in target_dir so that the cache can store them. The
+         dune cache doesn't support symlinks, so we replace them with hardlinks
+         to their targets. *)
+      let* () =
+        Async.async (fun () ->
+          if Path.Untracked.exists (Path.build target_dir)
+          then resolve_symlinks_in (Path.Build.to_string target_dir))
+      in
```
Comment (rgrinberg) [inline src/dune_rules/pkg_rules.ml]: I think you should combine this `Async.async` call with the one below it. There's a bunch of overhead for using the thread pool, so if it's to be used, the work should be batched as much as possible.

File: test/blackbox-tests/test-cases/pkg/install-symlinks/symlink-outside-target.t [hunk 3]
```diff
@@ -0,0 +1,55 @@
+Test symlinks pointing to files outside the target directory.
+
+CR-someday Alizter: This behavior is questionable. We convert the symlink to a
+hardlink pointing to an external file, which ties the build output to external
+content. If the external file changes, the target would be stale. Consider
+whether we should leave such symlinks as-is, copy the file, or error instead.
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/pkg/install-symlinks/symlink-outside-target.t:6]: This won't affect the cache, since we reject all symlinks there, but I am still unsure if this is good to accpt. At the end of the day, it probably won't matter much.

**Review by rgrinberg — APPROVED** [review]
